[RFC][DNM] drivers: adc: new API proposal

anangl opened the pull request:
This is an initial proposal of a new API for the ADC driver, so that we could start discussion and conclude what is actually needed in the API, and what will be convenient to use.
I tried to keep similarity with the new SPI API (#5921).

A few questions/doubts that I have:
1. Is there a need to use a bigger type than `s16_t` for sample values?
2. How can gain/prescaling settings be effectively passed to the driver? I suppose that these settings will vary significantly between different controllers.
3. Are there any other important configuration settings (of the ADC as a whole, or particular channels) that I haven't included yet?

@nashif @carlescufi @tbursztyka @MaureenHelm @andrewboie @galak

anangl committed:
drivers: adc: new API proposal

This is a proposal of a new API for ADC driver to address reported
issues (#3980) and clarify confusing things, like the buffer length.
Asynchronous requests and channel configuration were added as well.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
-	/* Pointer to a sequence entry array. */
-	struct adc_seq_entry *entries;
+struct adc_channel_cfg {
+	u8_t input_positive;
```
some description?

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
+	u8_t resolution;
+	/** Oversampling setting.
+	 *  Each sample is averaged from 2^oversampling results. */
+	u8_t oversampling;
```
so 0 would mean bare result?

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
+struct adc_sampling {
+	/** Clock ticks delay before starting the sampling (counted from the
+	 *  start of the previous sampling in the sequence, if any). */
+	u32_t delay;
```
ticks is hw dependent. Wouldn't it be better to work on µs or ms? (Well, µs is only supported through busy wait loop)

MaureenHelm replied:
I'm torn on this one. It's a pain to do the conversion in the driver and often inefficient due to divide operations, but I can see the value in having common units like us or ms for the application.

tbursztyka reviewed (commented):
samples value size will depend on resolution, so 8, 16, 24, 32, 64...?
Instead, I would make in void *, and depending on the resolution, the driver would fill it in accordingly.
Only thing would be that driver would check the buffer length vs resolution at first.
Just thinking out loud, I don't have much experience in ADCs (besides the awful adc_ti_adc108s102 I did back in the days)

galak commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
- * The routine returns once the ADC completes the read sequence.
- * The sample data can be retrieved from the memory buffers in
- * the sequence table structure.
+ * @brief Set an asynchronous read request.
```
This comment seems incorrect.

galak replied:
We should document the following - is it required to call adc_channel_setup before adc_read*?

galak reviewed (changes requested):
I get why adc_channel_setup() takes a dev pointer and the others dont, but the inconsistency is unfortunate.

galak commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
+
+static inline
+int _impl_adc_channel_setup(struct device *dev,
+			    u8_t channel_id,
```
any reason channel_id isn't in adc_channel_cfg?

vaussard commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
+	u8_t input_negative;
+	/**
+	 *  A bit field with the following parts:
+	 *  [5 :  0] - acquisition time, in microseconds
```
6 bits for the acquisition time looks a bit short, as 63 microseconds may be too small if you sample a high-impedance source or your ADC runs on a very slow clock.
On the other hand, 1us is rather high if you go high speed. For the STM32L4A6, I can setup Tacq as low as 30 nanoseconds.
So maybe we need a more flexible approach to specify the conversion time?

vaussard commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
-	u32_t buffer_length;
+	/** Pointer to a buffer where the samples are to be written.
+	 *  Each sample value is sign-extended to 16 bits. */
+	s16_t *buffer;
```
s16_t is not enough, I used external 24-bit ADC in the past. And you need more bits when doing oversampling.

MaureenHelm replied:
Agreed

vaussard commented:
A missing feature is the ability to setup the trigger source. Here a software trigger is implied, but we have to use hardware triggers sometime. I do not see a straightforward way to setup it as it is usually hardware dependant, but maybe some kind of driver-specific API (like uart_drv_cmd() or Linux ioctl) ?

anangl commented:
I will be on vacation all next week. I'll address all the comments (already existing ones and any new that may appear, and are welcome) when I'm back.

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
-
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+struct adc_sampling {
```
It would be more consistent with spi to separate them

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
+	 *  [6]      - mode, 0: single-ended (only positive input is specified)
+	 *                   1: differential (both inputs need to be specified)
+	 *  [7 :  ?] - gain/prescaling setting
+	 *  anything else ?
```
Do these things ever change on a per-channel basis? They don't on NXP devices

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,152 @@ extern "C" {
-
-	/** Pointer to the read routine. */
-	int (*read)(struct device *dev, struct adc_seq_table *seq_table);
+	int (*read)(const struct adc_config *config,
```
Please add the dev pointer argument back and remove it from the adc_config struct

MaureenHelm reviewed (commented):
> A missing feature is the ability to setup the trigger source. Here a software trigger is implied, but we have to use hardware triggers sometime. I do not see a straightforward way to setup it as it is usually hardware dependant,

Agreed

anangl commented:
Tried to address gathered feedback, to some extent at least.
Please consider it still as a work in progress, more a base for discussion than a real code.

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+				       struct k_poll_signal *async)
 {
-	const struct adc_driver_api *api = dev->driver_api;
+	const struct adc_driver_api *api = config->dev->driver_api;
```
tiny left over from previous PR push

anangl replied:
Right. I haven't tried asynchronous part yet. That's why I haven't got any compile error for this.

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
-
-	/** Pointer to the read routine. */
-	int (*read)(struct device *dev, struct adc_seq_table *seq_table);
+	int (*read)(struct device *dev,
```
to be consistent  with other APIs, create function typedefs for the functions first

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+	 * Each sample is averaged from 2^oversampling results.
+	 */
+	u8_t oversampling;
+
```
I would reorder the attributes so these 2 u8_t would be at the end of the struct.

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+
+struct adc_sequence_continuation {
+	/** Number of consecutive samplings to perform. */
+	u8_t sampling_count;
```
I would move this attribute at the end of the struct

tbursztyka commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+	/** ADC reference selection. */
+	enum adc_reference  reference;
+	/** Channel voltage prescaling selection. */
+	enum adc_prescaling prescaling;
```
move these 2 attributes at the beginning of the struct

tbursztyka reviewed (commented):
tiny changes requested. (mostly on struct alignment)

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+	/**
+	 * Channels to be used for getting consecutive samples.
```
Do they need to be unique? (no duplicates)

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+	/**
+	 * Number of samples to read in the sampling.
+	 */
+	u8_t sample_count;
```
Maybe this should be channel_count? It's a little confusing to have "samplings" and "samples" be different things

anangl replied:
I simplified the structure and used a bit-mask for specifying the channels to be included in the sampling. The previous proposal with the `channel_ids` pointer and `sample_count` was too confusing and probably troublesome in implementation (regarding duplicates, as you pointed out in the comment above, and order of the channels).

My idea was to define a sampling as reading of samples from one or more channels. Such samplings can be then arranged in sequences with optional triggers. I added more comments in the header file and changed some names to hopefully present the idea clearer. I am open to change the terminology to a better one. English is not my native language so the names I propose may not be the best (or even good enough) ones.

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+{
+	ADC_TRIGGER_DELAY_TICKS,
+	ADC_TRIGGER_DELAY_US,
+	ADC_TRIGGER_DELAY_MS,
```
Are the triggers one-off or recurring (such as from a hardware timer)? Can you explain the difference between the `DELAY` triggers and the `SAMPLE_RATE` trigger?

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+	/** Number of consecutive samplings to perform. */
+	u8_t sampling_count;
+
+	/** Type of trigger that starts each sampling. */
```
A sampling is a group of channels?

anangl replied:
Yes. I added better, hopefully, descriptions of the structure and its fields.

MaureenHelm commented on include/adc.h:
```diff
@@ -27,119 +27,225 @@ extern "C" {
+enum adc_prescaling
+{
+	ADC_PRESCALING_1_6,
+	ADC_PRESCALING_1_5,
```
Are these numbers non-integer gain factors? We will need more values in the enum, as the nxp adc supports gain factors of {1,2,4,8,16,32,64}

anangl replied:
Yes, these are fractional factors. I added proper comments, and renamed "prescaling" to "gain".

anangl commented:
Updated the proposal according to the recent suggestions. Added more descriptions to make things clearer.

anangl committed:
tests: drivers: Update adc_api test to check the new API

Updated API calls and added a few new test cases so that the new
possibilities of the API could be presented and checked.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
I updated the proposal once more. I got rid of the `adc_sampling` structure - its fields became members of the `adc_sequence` structure. I think this way it will be less confusing and simpler for example for single channel samplings, and the possibility of having a sequence of samplings each performed for a different sets of selected channels, which is no longer present, was actually not that much usable.
I added also actions that the application can pass from the callback to driver to alter the performed sequence.
And finally, I updated the adc_api test to show my vision of the usage of the proposed API.

tbursztyka commented:
Seems fine to me.
You seem to have forgotten the drivers update? (to get your test running)

anangl commented:
Replaced by #7691.
